WikiWatershed/model-my-watershed
# Combine PA Municipalities with Urbanization Layer

## Overview

Users would like to be able to view the PA Urbanization Layer and Municipalities in combination with hydrological boundaries such as HUC-12s. This is currently not possible since only one boundary may be active at a time.

To allow this, we combine the PA Municipalities layer with the Urbanization Layer into a new PA Urbanization with Municipalities coverage layer, which can be combined with any boundary layer.

We also develop a framework where any number of layers can be combined for any coverage layer going forward.

Connects #2769 

### Demo

![2018-06-26 14 36 05](https://user-images.githubusercontent.com/1430060/41932360-d147f85a-794e-11e8-9073-c8491dafc75b.gif)

### Notes

I tried adding an endpoint in Windshaft that did this, but could not figure out the SQL syntax for it. Also considered making a database view that had combined results, but thought that could be very expensive in terms of processing. Ended up going with a Leaflet based solution.

Upon deployment, we will need to clear the tile cache for this layer.

## Testing Instructions

* Check out this branch and `bundle`
* Restart the tiler:

    ```bash
    $ vagrant ssh tiler -c 'sudo service mmw-tiler restart'
    ```
* Go to [:8000/](http://localhost:8000/) and switch to the "Coverage" tab in the layer selector
* Select the new "PA Urbanization with Municipalities" layer. Ensure you see a gray overlay in urban areas and a dark green outline demarcating municipal boundaries.
  - You may have to open the network tab in dev tools and select "Disable Cache" to get the new tiles in case the old ones are cached in the browser
* Ensure changing the opacity works for both the grey and the dark green.
* Ensure zooming and panning loads tiles for both layers (albeit with a little delay, but always in the correct z-order).
* Enable a boundary layer, such as HUC-12. Ensure it renders atop the coverage layer and its styling is distinct from the municipal boundary.
* Enable a stream layer. Ensure it renders atop the coverage layer.
